grails-app/controllers/au/org/ala/ecodata/CommentController.groovy

```diff
@@ -0,0 +1,161 @@
+        Integer total = 0;
+
+        if( !entityId || !entityType){
+            response.sendError(400, 'Insufficient parameters provided. Missing either entityId or entityType')
+        } else {
+//           todo: is there a better way to do count. is the below method potentially slow?
+            total = Comment.findAllWhere([ 'entityId':entityId, 'entityType':entityType, parent: null]).size();
+            Comment.findAllWhere(['entityId':entityId, 'entityType':entityType, parent: null], [sort:sort,order:order,offset:offset,max:max]).each {
+                Map mapOfProperties = outputService.getCommentProperties(it)
+                comments.add(mapOfProperties)
+            }
+
+            comments.each { comment ->
+                if(comment.children?.size()){
+                    outputService.sortCommentChildren(comment.children, sortOrder);
+                }
+            }
+
+            render text:  [total: total, items: comments] as JSON, contentType: 'application/json'
+        }
+    }
+
+
+    @RequireApiKey
+    def create(){
+        def jsonSlurper = new JsonSlurper()
+        def json = jsonSlurper.parse(request.getReader())
+        if (!json.userId){
+            response.sendError(400, 'Missing userId')
+        } else if (!json.entityId || !json.entityType){
+            response.sendError(400, 'Missing entityId and/or entityType')
+        } else if ( !json.text ){
+            response.sendError(400, 'Missing text');
+        } else {
+            Comment l = new Comment(json)
+            Comment parent;
+            if(l.dateCreated == null){
+                l.dateCreated = new Date();
+            }
+
+            Comment comment = l.save(true)
+
+            if(json.parent != null){
+                parent = Comment.get(json.parent);
+                if(parent){
+                    comment.parent = parent;
+                    parent.children.add(comment);
+                    parent.save(true);
+                    comment.save(true);
+                }
+            }
+
+
+
+            response.addHeader("content-location", grailsApplication.config.grails.serverURL + "/comment/" + comment.getId().toString())
+            response.addHeader("location", grailsApplication.config.grails.serverURL + "/comment/" + comment.getId().toString())
+            response.addHeader("entityId", comment.getId().toString())
+            response.setContentType("application/json")
+            def model = outputService.getCommentProperties(comment)
+            render model as JSON
+        }
+
+    }
+
+    @RequireApiKey
+    def update(){
+        def jsonSlurper = new JsonSlurper()
+        def json = jsonSlurper.parse(request.getReader())
+        if(!json.id){
+            response.sendError(400, "Missing id");
+        } else if ( !json.text ){
+            response.sendError(400, 'Missing text');
+        } else {
+            Map result
+            Comment c = Comment.get(json.id);
+            if(c){
```

Comment:
Why use this syntax for accessing properties instead of the stardard x.y? Using strings like this makes refactoring and finding usages of properties harder.

Also, single character names again.